src/input/keyboard/xkb_config.rs

```diff
@@ -27,19 +28,34 @@ pub struct XkbConfig<'a> {
     /// preferences, like which key combinations are used for switching layouts, or which key is the
     /// Compose key.
     pub options: Option<String>,
+    /// Path to a file from which the keymap can be compiled. Allows the user to provide a
+    /// stand-alone keymap file that will be used instead of a system keymap.
+    pub file: Option<String>,
 }
 
 impl<'a> XkbConfig<'a> {
     pub(crate) fn compile_keymap(&self, context: &xkb::Context) -> Result<xkb::Keymap, ()> {
-        xkb::Keymap::new_from_names(
-            context,
-            self.rules,
-            self.model,
-            self.layout,
-            self.variant,
-            self.options.clone(),
-            xkb::KEYMAP_COMPILE_NO_FLAGS,
-        )
-        .ok_or(())
+        match &self.file {
+            Some(f) => {
+                let mut file = File::open(f).map_err(|_| ())?;
```

Comment:
It's probably not good to have a `File::open` where errors are just mapped to `()`, without at least logging something. That could be confusing if the file simply doesn't exist.